src/io/udp_frame.rs

```diff
@@ -0,0 +1,251 @@
+use std::io;
+use std::net::SocketAddr;
+use net::UdpSocket;
+use futures::{Async, Poll, Stream, Sink, StartSend, AsyncSink};
+use futures::sync::BiLock;
+
+/// Encoding of frames via buffers.
+///
+/// This trait is used when constructing an instance of `FramedUdp`. It provides
+/// one type: `Out` for encoding outgoing frames according to a protocol.
+///
+/// Because UDP is a connectionless protocol, the encode method will also be
+/// responsible for determining the remote host to which the datagram should be
+/// sent
+///
+/// The trait itself is implemented on a type that can track state for decoding
+/// or encoding, which is particularly useful for streaming parsers. In many
+/// cases, though, this type will simply be a unit struct (e.g. `struct
+/// HttpCodec`).
+pub trait CodecUdp {
+
+    /// The type of frames to be encoded.
+    type Out;
+    
+    /// The type of decoded frames.
+    type In;
+
+
+    /// Encodes a frame into the buffer provided.
+    ///
+    /// This method will encode `msg` into the byte buffer provided by `buf`.
+    /// The `buf` provided is an internal buffer of the `Framed` instance and
+    /// will be written out when possible. 
+    ///
+    /// The encode method also determines the destination to which the buffer should
+    /// be directed, which will be returned as a SocketAddr;
+    fn encode(&mut self, msg: &Self::Out, buf: &mut Vec<u8>) -> SocketAddr;
+    
+    /// Attempts to decode a frame from the provided buffer of bytes.
+    ///
+    /// This method is called by `FramedUdp` on a single datagram which has been
+    /// read from a socket. 
+    ///
+    /// It is required that the decode method empty the read buffer in every call to
+    /// decode, as the next poll_read that occurs will write the next datagram
+    /// into the buffer, without regard for what is already there. 
+    ///
+    /// If the bytes look valid, but a frame isn't fully available yet, then
+    /// `Ok(None)` is returned. This indicates to the `Framed` instance that
+    /// it needs to read some more bytes before calling this method again.
+    /// In such a case, it is decode's responsibility to copy the data
+    /// into their own internal buffer for future use.
+    ///
+    /// Finally, if the bytes in the buffer are malformed then an error is
+    /// returned indicating why. This informs `Framed` that the stream is now
+    /// corrupt and should be terminated.
+    ///
+    fn decode(&mut self, src: &SocketAddr, buf: &[u8]) -> Result<Option<Self::In>, io::Error>;
```

Comment:
I think the receive -> decode behavior in datagrams should be pretty much the same for streams. 

If, in the case that some protocol frame is divided up into N chunks, when we receive chunk < N,
it seems like it would be good to indicate that we did receive data, and we are accumulating it in the codec until we can return a complete frame.  

However, in either case, it looks like the behavior to the consumer (Stream) would be the same, which is
`return Ok(Async::NotReady)`

The implementor of the codec would be responsible to associate the incomplete message with the inbound/outbound socketaddr.  

Were you thinking that the handling of chunked messages should be done by the user of FramedUdp rather than the implementor of Codec?